### PR TITLE
:sparkles: add work source client

### PR DESCRIPTION
**Pull request by skeeey**
## Related issue(s)

Fixes #

### CODE REVIEW

File: pkg/cloudevents/generic/baseclient.go [hunk 1]
```diff
@@ -71,6 +72,7 @@ func (c *baseClient) connect(ctx context.Context) error {
 				klog.V(4).Infof("the cloudevents client is reconnected")
 				c.resetClient(cloudEventsClient)
 				c.sendReceiverSignal(restartReceiverSignal)
+				c.sendReconnectedSignal()
```
Comment (qiujian16) [inline pkg/cloudevents/generic/baseclient.go:75]: will we keeps on sending reconnect signal here?

File: pkg/cloudevents/work/clientbuilder.go [hunk 2]
```diff
@@ -58,6 +70,7 @@ type ClientHolderBuilder struct {
 // Available configurations:
 //   - Kubeconfig (*rest.Config): builds a manifestwork client with kubeconfig
 //   - MQTTOptions (*mqtt.MQTTOptions): builds a manifestwork client based on cloudevents with MQTT
+//   - GRPCOptions (*mqtt.GRPCOptions): builds a manifestwork client based on cloudevents with GRPC
```
Comment (morvencao) [inline pkg/cloudevents/work/clientbuilder.go:73]: s/mqtt.GRPCOptions/grpc.GRPCOptions/

File: pkg/cloudevents/generic/baseclient.go (hunk 1 above)
Comment (skeeey) [reply pkg/cloudevents/generic/baseclient.go:75]: No, only when the reconnection is happened and client is already reconnected

File: pkg/cloudevents/generic/interface.go [hunk 3]
```diff
@@ -67,4 +67,8 @@ type CloudEventsClient[T ResourceObject] interface {
 	// Subscribe the resources status/spec event to the broker to receive the resources status/spec and use
 	// ResourceHandler to handle them.
 	Subscribe(ctx context.Context, handlers ...ResourceHandler[T])
+
+	// ReconnectedChan returns a chan which indicates the source/agent client is reconnected.
+	// The source/agent client callers should consider sending a resync request when receiving this signal.
+	ReconnectedChan() <-chan struct{}
```
[the remark this answers is not in the document]
Comment (skeeey) [reply pkg/cloudevents/generic/interface.go:73]: I think this is just a signal to tell the source/agent client callers the source/agent client is reconnected without additional information, so I used a dumb struct here

File: pkg/cloudevents/work/source/client/manifestwork.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+	}
+
+	if !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	eventDataType, err := types.ParseCloudEventsDataType(manifestWork.Annotations[common.CloudEventsDataTypeAnnotationKey])
+	if err != nil {
+		return nil, err
+	}
```
Comment (morvencao) [inline pkg/cloudevents/work/source/client/manifestwork.go:66]: does the manifestwork created from manifestworkreplicaset in source contains this annotation in that stage?

File: pkg/cloudevents/work/common/common.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+const (
+	// CloudEventsDataTypeAnnotationKey is the key of the cloudevents data type annotation.
+	CloudEventsDataTypeAnnotationKey = "cloudevents.open-cluster-management.io/datatype"
+
+	// CloudEventsGenerationAnnotationKey is the key of the manifestwork generation annotation.
+	CloudEventsGenerationAnnotationKey = "cloudevents.open-cluster-management.io/generation"
+)
+
+// CloudEventsOriginalSourceLabelKey is the key of the cloudevents original source annotation.
```
[the remark this answers is not in the document]
Comment (skeeey) [reply pkg/cloudevents/work/common/common.go]: updated

**Review by qiujian16 — COMMENTED** [review]
/approve

**Review by morvencao — COMMENTED** [review]
/lgtm

**Pull request merged**